File: c-implementation/print.c
#include "mem.h"

// print :: a -> ()
void print(cell x) {
  if (x == Nil()) {
    printf("()");
    return;
  }

  if (x->type == CONS) {
    printf("(");
    char space = 0;
    while (x != Nil() && x->type == CONS) {
      if (space) printf(" ");
      print(x->data.c.car);
      space = 1;
      x = x->data.c.cdr;
    }

    if (x != Nil()) {
      printf(" . ");
      print(x);
    }
    
    printf(")");
  } else if (x->type == INT) {
    printf("%i", x->data.i);
  } else if (x->type == SYMBOL) {
    printf("%s", x->data.s);
  } else if (x->type == CORE) {
    printf("[CORE]");
  } else if (x->type == FN) {
    printf("[FN ");
    print(x->data.c.car);
    printf("]");
  } else if (x->type == MACRO) {
    printf("[MACRO ");
    print(x->data.c.car);
    printf("]");
  }
}

File: src/print.js
const { car, cdr, EMPTY, isCons } = require("./list.js")

const prn = x => {
  if (x === undefined || x === null) return String(x)

  if (typeof x === "boolean" || typeof x === "number") return String(x)
  if (typeof x === "symbol") return Symbol.keyFor(x)
  if (typeof x === "string") return `"${x}"`

  if (isCons(x)) {
    const r = []
    while (x !== EMPTY) {
      if (isCons(x)) {
        r.push(prn(car(x)))
        x = cdr(x)
      } else {
        r.push(".")
        r.push(prn(x))
        x = EMPTY
      }
    }

    return `(${r.join(" ")})`
  }

  if (x instanceof Array) {
    return `[${x.map(prn).join(" ")}]`
  }

  if (x && x[Symbol.for("ISeq")]) {
    return `(${[...x].map(prn).join(" ")})`
  }

  if (typeof x === "function") {
    return x.macro ? "[macro]" : "[procedure]"
  }

  return `{${Object.entries(x)
    .map(([key, val]) => `"${key}" ${prn(val)}`)
    .join(" ")}}`
}

module.exports = prn
